File: reference/bindings/v0.2/typescript/src/boundary-crossing.ts
export type BoundaryType =
  | "EXECUTION" | "DISCLOSURE" | "MANDATE" | "RESOURCE" | "EVALUATION";

export type BoundaryState =
  | "NO_CROSSING" | "AUTHORIZED" | "CONDITIONALLY_AUTHORIZED"
  | "UNAUTHORIZED" | "INDETERMINATE" | "STALE" | "REVOKED";

export type BoundaryResponseFloor =
  | "NONE" | "MODIFY" | "DEFER" | "STEP_UP" | "DENY" | "HALT";

export interface ArtifactBinding {
  ref: string;
  digest: string;
}

export interface BoundaryAssessmentBinding {
  assessment_ref: string;
  assessment_digest: string;
}

export interface BoundaryRequirementSet {
  required_types: BoundaryType[];
  policy_ref: string;
  policy_digest: string;
  fail_closed: true;
}

export interface BoundaryCrossing {
  crossing_id: string;
  boundary_type: BoundaryType;
  crossing_detected: boolean;
  prior_state_digest: string;
  proposed_state_digest: string;
  authority_requirement_ref: string;
  authority_binding?: ArtifactBinding;
  policy_binding: ArtifactBinding;
  evidence_binding: ArtifactBinding;
  resource_reservation_binding?: ArtifactBinding;
  evaluation_provenance_binding?: ArtifactBinding;
  details_digest: string;
  state: BoundaryState;
  required_response_floor: BoundaryResponseFloor;
  reason_codes: string[];
  observed_at: string;
  valid_until: string;
}

export class BoundaryCrossingAssessment {
  schema_version!: "racs.boundary-crossing-assessment.v0.2";
  assessment_id!: string;
  action_id!: string;
  action_envelope_digest!: string;
  tenant_id!: string;
  assessor_id!: string;
  assessor_version!: string;
  requirement_policy_ref!: string;
  requirement_policy_digest!: string;
  crossings!: BoundaryCrossing[];
  aggregate_state!: BoundaryState;
  required_response_floor!: BoundaryResponseFloor;
  reason_codes!: string[];
  assessed_at!: string;
  valid_until!: string;
  revocation_registry_ref!: string;
}

const BOUNDARY_ORDER: Record<BoundaryType, number> = {
  EXECUTION: 0, DISCLOSURE: 1, MANDATE: 2, RESOURCE: 3, EVALUATION: 4,
};
const STATE_RANK: Record<BoundaryState, number> = {
  NO_CROSSING: 0, AUTHORIZED: 1, CONDITIONALLY_AUTHORIZED: 2,
  INDETERMINATE: 3, UNAUTHORIZED: 4, STALE: 5, REVOKED: 6,
};
const RESPONSE_RANK: Record<BoundaryResponseFloor, number> = {
  NONE: 0, MODIFY: 1, DEFER: 2, STEP_UP: 3, DENY: 4, HALT: 5,
};

function validTime(value: string): number {
  const parsed = Date.parse(value);
  if (Number.isNaN(parsed) || !/(Z|[+-]\d\d:\d\d)$/.test(value)) {
    throw new Error("timestamp must include timezone");
  }
  return parsed;
}

function uniqueSorted(values: string[]): boolean {
  return new Set(values).size === values.length &&
    values.every((value, index) => index === 0 || values[index - 1] <= value);
}

function crossingSemanticError(crossing: BoundaryCrossing): string | null {
  const observed = validTime(crossing.observed_at);
  const until = validTime(crossing.valid_until);
  if (until <= observed) return "crossing valid_until must be after observed_at";
  if (!uniqueSorted(crossing.reason_codes)) return "reason_codes must be unique and sorted";

  const changed = crossing.prior_state_digest !== crossing.proposed_state_digest;
  if (crossing.crossing_detected !== changed) {
    return "crossing_detected must equal state-digest change";
  }
  if (!crossing.crossing_detected) {
    if (crossing.state !== "NO_CROSSING") return "non-crossing must use NO_CROSSING";
    if (crossing.required_response_floor !== "NONE") return "non-crossing must use NONE response";
    if (crossing.reason_codes.length > 0) return "non-crossing cannot carry reason codes";
    return null;
  }
  if (crossing.state === "NO_CROSSING") return "detected crossing cannot use NO_CROSSING";
  if (
    (crossing.state === "AUTHORIZED" || crossing.state === "CONDITIONALLY_AUTHORIZED") &&
    crossing.authority_binding === undefined
  ) {
    return "authorized crossing requires authority_binding";
  }

  const minimumByState: Record<Exclude<BoundaryState, "NO_CROSSING">, BoundaryResponseFloor> = {
    AUTHORIZED: "NONE",
    CONDITIONALLY_AUTHORIZED: "MODIFY",
    INDETERMINATE: "DEFER",
    UNAUTHORIZED: "DENY",
    STALE: "DEFER",
    REVOKED: "DENY",
  };
  const minimum = minimumByState[crossing.state as Exclude<BoundaryState, "NO_CROSSING">];
  if (RESPONSE_RANK[crossing.required_response_floor] < RESPONSE_RANK[minimum]) {
    return `${crossing.state} requires ${minimum} or stronger response`;
  }
  if (crossing.state === "AUTHORIZED" && crossing.required_response_floor !== "NONE") {
    return "AUTHORIZED crossing must use NONE response";
  }
  if (
    crossing.state === "CONDITIONALLY_AUTHORIZED" &&
    crossing.required_response_floor !== "MODIFY"
  ) {
    return "CONDITIONALLY_AUTHORIZED crossing must use MODIFY response";
  }

  const reasons = new Set(crossing.reason_codes);
  if (reasons.has("TECHNICAL_ACCESS_ONLY")) {
    if (crossing.state !== "UNAUTHORIZED") {
      return "technical access alone cannot authorize execution";
    }
    if (
      crossing.required_response_floor !== "DENY" &&
      crossing.required_response_floor !== "HALT"
    ) {
      return "technical access alone requires DENY or HALT";
    }
  }
  if (reasons.has("UNAUTHORIZED_DISCOVERABILITY") && crossing.state !== "UNAUTHORIZED") {
    return "unauthorized discoverability must be UNAUTHORIZED";
  }
  if (reasons.has("RESOURCE_LIMIT_EXCEEDED") && crossing.state !== "UNAUTHORIZED") {
    return "resource limit exceeded must be UNAUTHORIZED";
  }
  if (
    crossing.boundary_type === "RESOURCE" &&
    (crossing.state === "AUTHORIZED" || crossing.state === "CONDITIONALLY_AUTHORIZED") &&
    crossing.resource_reservation_binding === undefined
  ) {
    return "authorized resource crossing requires reservation binding";
  }
  if (
    crossing.boundary_type === "EVALUATION" &&
    crossing.evaluation_provenance_binding === undefined
  ) {
    return "evaluation crossing requires provenance binding";
  }
  return null;
}

export function boundaryAssessmentSemanticError(
  assessment: BoundaryCrossingAssessment,
): string | null {
  const assessed = validTime(assessment.assessed_at);
  const validUntil = validTime(assessment.valid_until);
  if (validUntil <= assessed) return "assessment valid_until must be after assessed_at";

  const types = assessment.crossings.map((item) => item.boundary_type);
  if (new Set(types).size !== types.length) return "assessment cannot repeat boundary types";
  if (!types.every((type, index) => index === 0 ||
      BOUNDARY_ORDER[types[index - 1]] <= BOUNDARY_ORDER[type])) {
    return "crossings must use canonical boundary order";
  }
  if (!types.includes("EXECUTION")) return "assessment must include EXECUTION boundary";

  const ids = assessment.crossings.map((item) => item.crossing_id);
  if (new Set(ids).size !== ids.length) return "crossing_id values must be unique";

  for (const crossing of assessment.crossings) {
    const error = crossingSemanticError(crossing);
    if (error) return error;
    if (crossing.policy_binding.ref !== assessment.requirement_policy_ref) {
      return "crossing policy ref must match requirement policy";
    }
    if (crossing.policy_binding.digest !== assessment.requirement_policy_digest) {
      return "crossing policy digest must match requirement policy";
    }
    if (validTime(crossing.observed_at) > assessed) {
      return "crossing cannot be observed after assessment";
    }
    if (validTime(crossing.valid_until) < validUntil) {
      return "assessment cannot outlive crossing evidence";
    }
  }

  const expectedState = assessment.crossings.reduce(
    (left, right) => STATE_RANK[left.state] >= STATE_RANK[right.state] ? left : right,
  ).state;
  const expectedResponse = assessment.crossings.reduce(
    (left, right) =>
      RESPONSE_RANK[left.required_response_floor] >= RESPONSE_RANK[right.required_response_floor]
        ? left : right,
  ).required_response_floor;
  const expectedReasons = [...new Set(
    assessment.crossings.flatMap((item) => item.reason_codes),
  )].sort();

  if (assessment.aggregate_state !== expectedState) {
    return "aggregate_state does not match crossings";
  }
  if (assessment.required_response_floor !== expectedResponse) {
    return "required_response_floor does not match crossings";
  }
  if (JSON.stringify(assessment.reason_codes) !== JSON.stringify(expectedReasons)) {
    return "assessment reason_codes must equal sorted crossing union";
  }
  return null;
}

export function responseFloorSatisfied(
  responseFloor: BoundaryResponseFloor,
  decision: string,
): boolean {
  const allowed: Record<BoundaryResponseFloor, Set<string>> = {
    NONE: new Set(["ALLOW", "MODIFY", "DEFER", "STEP_UP", "DENY", "HALT"]),
    MODIFY: new Set(["MODIFY", "DEFER", "STEP_UP", "DENY", "HALT"]),
    DEFER: new Set(["DEFER", "STEP_UP", "DENY", "HALT"]),
    STEP_UP: new Set(["STEP_UP", "DENY", "HALT"]),
    DENY: new Set(["DENY", "HALT"]),
    HALT: new Set(["HALT"]),
  };
  return allowed[responseFloor].has(decision);
}

File: reference/bindings/v0.2/rust/src/boundary_crossing.rs
use chrono::DateTime;
use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use std::error::Error;

use crate::{canonical_string, sha256_digest};

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum BoundaryType {
    Execution,
    Disclosure,
    Mandate,
    Resource,
    Evaluation,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum BoundaryState {
    NoCrossing,
    Authorized,
    ConditionallyAuthorized,
    Unauthorized,
    Indeterminate,
    Stale,
    Revoked,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum BoundaryResponseFloor {
    None,
    Modify,
    Defer,
    StepUp,
    Deny,
    Halt,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct ArtifactBinding {
    pub r#ref: String,
    pub digest: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct BoundaryAssessmentBinding {
    pub assessment_ref: String,
    pub assessment_digest: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct BoundaryRequirementSet {
    pub required_types: Vec<BoundaryType>,
    pub policy_ref: String,
    pub policy_digest: String,
    pub fail_closed: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct BoundaryCrossing {
    pub crossing_id: String,
    pub boundary_type: BoundaryType,
    pub crossing_detected: bool,
    pub prior_state_digest: String,
    pub proposed_state_digest: String,
    pub authority_requirement_ref: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub authority_binding: Option<ArtifactBinding>,
    pub policy_binding: ArtifactBinding,
    pub evidence_binding: ArtifactBinding,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub resource_reservation_binding: Option<ArtifactBinding>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub evaluation_provenance_binding: Option<ArtifactBinding>,
    pub details_digest: String,
    pub state: BoundaryState,
    pub required_response_floor: BoundaryResponseFloor,
    #[serde(default)]
    pub reason_codes: Vec<String>,
    pub observed_at: String,
    pub valid_until: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct BoundaryCrossingAssessment {
    pub schema_version: String,
    pub assessment_id: String,
    pub action_id: String,
    pub action_envelope_digest: String,
    pub tenant_id: String,
    pub assessor_id: String,
    pub assessor_version: String,
    pub requirement_policy_ref: String,
    pub requirement_policy_digest: String,
    pub crossings: Vec<BoundaryCrossing>,
    pub aggregate_state: BoundaryState,
    pub required_response_floor: BoundaryResponseFloor,
    #[serde(default)]
    pub reason_codes: Vec<String>,
    pub assessed_at: String,
    pub valid_until: String,
    pub revocation_registry_ref: String,
}

fn parse_time(value: &str) -> Result<DateTime<chrono::FixedOffset>, String> {
    if !(value.ends_with('Z')
        || value
            .get(value.len().saturating_sub(6)..)
            .map(|tail| {
                let bytes = tail.as_bytes();
                tail.len() == 6
                    && (bytes[0] == b'+' || bytes[0] == b'-')
                    && bytes[3] == b':'
            })
            .unwrap_or(false))
    {
        return Err("timestamp must include timezone".into());
    }
    DateTime::parse_from_rfc3339(&value.replace('Z', "+00:00"))
        .map_err(|_| "timestamp must include timezone".into())
}

fn boundary_rank(value: BoundaryType) -> u8 {
    match value {
        BoundaryType::Execution => 0,
        BoundaryType::Disclosure => 1,
        BoundaryType::Mandate => 2,
        BoundaryType::Resource => 3,
        BoundaryType::Evaluation => 4,
    }
}

fn state_rank(value: BoundaryState) -> u8 {
    match value {
        BoundaryState::NoCrossing => 0,
        BoundaryState::Authorized => 1,
        BoundaryState::ConditionallyAuthorized => 2,
        BoundaryState::Indeterminate => 3,
        BoundaryState::Unauthorized => 4,
        BoundaryState::Stale => 5,
        BoundaryState::Revoked => 6,
    }
}

fn response_rank(value: BoundaryResponseFloor) -> u8 {
    match value {
        BoundaryResponseFloor::None => 0,
        BoundaryResponseFloor::Modify => 1,
        BoundaryResponseFloor::Defer => 2,
        BoundaryResponseFloor::StepUp => 3,
        BoundaryResponseFloor::Deny => 4,
        BoundaryResponseFloor::Halt => 5,
    }
}

fn sorted_unique(values: &[String]) -> bool {
    let unique: HashSet<&String> = values.iter().collect();
    unique.len() == values.len() && values.windows(2).all(|pair| pair[0] <= pair[1])
}

impl BoundaryRequirementSet {
    pub fn validate_semantics(&self) -> Result<(), String> {
        if self.required_types.is_empty() {
            return Err("required_types must not be empty".into());
        }
        let unique: HashSet<BoundaryType> = self.required_types.iter().copied().collect();
        if unique.len() != self.required_types.len() {
            return Err("required_types must be unique".into());
        }
        if !self
            .required_types
            .windows(2)
            .all(|pair| boundary_rank(pair[0]) <= boundary_rank(pair[1]))
        {
            return Err("required_types must use canonical boundary order".into());
        }
        if !self.required_types.contains(&BoundaryType::Execution) {
            return Err("EXECUTION boundary is mandatory for every ActionEnvelope".into());
        }
        if !self.fail_closed {
            return Err("boundary requirements must be fail_closed".into());
        }
        Ok(())
    }
}

impl BoundaryCrossing {
    pub fn validate_semantics(&self) -> Result<(), String> {
        let observed = parse_time(&self.observed_at)?;
        let valid_until = parse_time(&self.valid_until)?;
        if valid_until <= observed {
            return Err("crossing valid_until must be after observed_at".into());
        }
        if !sorted_unique(&self.reason_codes) {
            return Err("reason_codes must be unique and sorted".into());
        }

        let changed = self.prior_state_digest != self.proposed_state_digest;
        if self.crossing_detected != changed {
            return Err("crossing_detected must equal state-digest change".into());
        }
        if !self.crossing_detected {
            if self.state != BoundaryState::NoCrossing {
                return Err("non-crossing must use NO_CROSSING".into());
            }
            if self.required_response_floor != BoundaryResponseFloor::None {
                return Err("non-crossing must use NONE response".into());
            }
            if !self.reason_codes.is_empty() {
                return Err("non-crossing cannot carry reason codes".into());
            }
            return Ok(());
        }

        if self.state == BoundaryState::NoCrossing {
            return Err("detected crossing cannot use NO_CROSSING".into());
        }
        if matches!(
            self.state,
            BoundaryState::Authorized | BoundaryState::ConditionallyAuthorized
        ) && self.authority_binding.is_none()
        {
            return Err("authorized crossing requires authority_binding".into());
        }

        let minimum = match self.state {
            BoundaryState::Authorized => BoundaryResponseFloor::None,
            BoundaryState::ConditionallyAuthorized => BoundaryResponseFloor::Modify,
            BoundaryState::Indeterminate => BoundaryResponseFloor::Defer,
            BoundaryState::Unauthorized => BoundaryResponseFloor::Deny,
            BoundaryState::Stale => BoundaryResponseFloor::Defer,
            BoundaryState::Revoked => BoundaryResponseFloor::Deny,
            BoundaryState::NoCrossing => BoundaryResponseFloor::None,
        };
        if response_rank(self.required_response_floor) < response_rank(minimum) {
            return Err("boundary response is weaker than required state minimum".into());
        }
        if self.state == BoundaryState::Authorized
            && self.required_response_floor != BoundaryResponseFloor::None
        {
            return Err("AUTHORIZED crossing must use NONE response".into());
        }
        if self.state == BoundaryState::ConditionallyAuthorized
            && self.required_response_floor != BoundaryResponseFloor::Modify
        {
            return Err("CONDITIONALLY_AUTHORIZED crossing must use MODIFY response".into());
        }

        let reasons: HashSet<&str> = self.reason_codes.iter().map(String::as_str).collect();
        if reasons.contains("TECHNICAL_ACCESS_ONLY") {
            if self.state != BoundaryState::Unauthorized {
                return Err("technical access alone cannot authorize execution".into());
            }
            if !matches!(
                self.required_response_floor,
                BoundaryResponseFloor::Deny | BoundaryResponseFloor::Halt
            ) {
                return Err("technical access alone requires DENY or HALT".into());
            }
        }
        if reasons.contains("UNAUTHORIZED_DISCOVERABILITY")
            && self.state != BoundaryState::Unauthorized
        {
            return Err("unauthorized discoverability must be UNAUTHORIZED".into());
        }
        if reasons.contains("RESOURCE_LIMIT_EXCEEDED")
            && self.state != BoundaryState::Unauthorized
        {
            return Err("resource limit exceeded must be UNAUTHORIZED".into());
        }
        if self.boundary_type == BoundaryType::Resource
            && matches!(
                self.state,
                BoundaryState::Authorized | BoundaryState::ConditionallyAuthorized
            )
            && self.resource_reservation_binding.is_none()
        {
            return Err("authorized resource crossing requires reservation binding".into());
        }
        if self.boundary_type == BoundaryType::Evaluation
            && self.evaluation_provenance_binding.is_none()
        {
            return Err("evaluation crossing requires provenance binding".into());
        }
        Ok(())
    }
}

impl BoundaryCrossingAssessment {
    pub fn canonical(&self) -> Result<String, Box<dyn Error>> {
        canonical_string(self)
    }

    pub fn digest(&self) -> Result<String, Box<dyn Error>> {
        sha256_digest(self)
    }

    pub fn validate_semantics(&self) -> Result<(), String> {
        if self.schema_version != "racs.boundary-crossing-assessment.v0.2" {
            return Err("invalid boundary assessment schema_version".into());
        }
        let assessed = parse_time(&self.assessed_at)?;
        let valid_until = parse_time(&self.valid_until)?;
        if valid_until <= assessed {
            return Err("assessment valid_until must be after assessed_at".into());
        }
        if self.crossings.is_empty() {
            return Err("assessment must include at least one crossing".into());
        }

        let types: Vec<BoundaryType> =
            self.crossings.iter().map(|item| item.boundary_type).collect();
        let unique_types: HashSet<BoundaryType> = types.iter().copied().collect();
        if unique_types.len() != types.len() {
            return Err("assessment cannot repeat boundary types".into());
        }
        if !types
            .windows(2)
            .all(|pair| boundary_rank(pair[0]) <= boundary_rank(pair[1]))
        {
            return Err("crossings must use canonical boundary order".into());
        }
        if !types.contains(&BoundaryType::Execution) {
            return Err("assessment must include EXECUTION boundary".into());
        }

        let ids: HashSet<&String> = self.crossings.iter().map(|item| &item.crossing_id).collect();
        if ids.len() != self.crossings.len() {
            return Err("crossing_id values must be unique".into());
        }

        for crossing in &self.crossings {
            crossing.validate_semantics()?;
            if crossing.policy_binding.r#ref != self.requirement_policy_ref {
                return Err("crossing policy ref must match requirement policy".into());
            }
            if crossing.policy_binding.digest != self.requirement_policy_digest {
                return Err("crossing policy digest must match requirement policy".into());
            }
            if parse_time(&crossing.observed_at)? > assessed {
                return Err("crossing cannot be observed after assessment".into());
            }
            if parse_time(&crossing.valid_until)? < valid_until {
                return Err("assessment cannot outlive crossing evidence".into());
            }
        }

        let expected_state = self
            .crossings
            .iter()
            .max_by_key(|item| state_rank(item.state))
            .map(|item| item.state)
            .ok_or_else(|| "assessment must include at least one crossing".to_string())?;
        let expected_response = self
            .crossings
            .iter()
            .max_by_key(|item| response_rank(item.required_response_floor))
            .map(|item| item.required_response_floor)
            .ok_or_else(|| "assessment must include at least one crossing".to_string())?;
        let mut expected_reasons: Vec<String> = self
            .crossings
            .iter()
            .flat_map(|item| item.reason_codes.iter().cloned())
            .collect::<HashSet<String>>()
            .into_iter()
            .collect();
        expected_reasons.sort();

        if self.aggregate_state != expected_state {
            return Err("aggregate_state does not match crossings".into());
        }
        if self.required_response_floor != expected_response {
            return Err("required_response_floor does not match crossings".into());
        }
        if self.reason_codes != expected_reasons {
            return Err("assessment reason_codes must equal sorted crossing union".into());
        }
        Ok(())
    }
}

pub fn response_floor_satisfied(
    response_floor: BoundaryResponseFloor,
    decision: crate::Decision,
) -> bool {
    match response_floor {
        BoundaryResponseFloor::None => true,
        BoundaryResponseFloor::Modify => !matches!(decision, crate::Decision::Allow),
        BoundaryResponseFloor::Defer => matches!(
            decision,
            crate::Decision::Defer
                | crate::Decision::StepUp
                | crate::Decision::Deny
                | crate::Decision::Halt
        ),
        BoundaryResponseFloor::StepUp => matches!(
            decision,
            crate::Decision::StepUp | crate::Decision::Deny | crate::Decision::Halt
        ),
        BoundaryResponseFloor::Deny => {
            matches!(decision, crate::Decision::Deny | crate::Decision::Halt)
        }
        BoundaryResponseFloor::Halt => matches!(decision, crate::Decision::Halt),
    }
}
